crates/driver/src/domain/competition/mod.rs

```diff
@@ -95,36 +97,100 @@ impl Competition {
-                })
-                .ok()
-        });
-
-        // Score the solutions.
-        let scores = join_all(settlements.map(|settlement| async move {
-            tracing::trace!(id = ?settlement.solution.id, "scoring");
-            (settlement.score(&self.eth, auction).await, settlement)
-        }))
-        .await;
-
-        // Filter out solutions which failed scoring.
-        let scores = scores
+        let mut settlements = settlements
             .into_iter()
-            .filter_map(|(result, settlement)| {
+            .filter_map(|(id, result)| {
                 result
                     .tap_err(|err| {
-                        tracing::info!(?err, id = ?settlement.solution.id, "discarding solution: failed scoring")
+                        tracing::info!(?err, ?id, "discarding solution: failed verification")
                     })
                     .ok()
-                    .map(|score| (score, settlement))
-            });
+            })
+            .collect_vec();
+
+        // TODO(#1483): parallelize this
+
+        // Merge the settlements in random order.
+        settlements.shuffle(&mut rand::thread_rng());
+        // The merged settlements in their final form.
+        let mut results = Vec::new();
+        while !settlements.is_empty() {
+            let settlement = settlements.pop().unwrap();
+            // Has [`settlement`] been merged into another settlement?
+            let mut merged = false;
+            // Try to merge [`settlement`] into some other settlement.
+            for other in settlements.iter_mut() {
+                match other
+                    .merge(
+                        &settlement,
+                        self.eth.contracts().settlement(),
+                        &self.eth,
+                        &self.simulator,
+                    )
+                    .await
+                {
+                    Ok(m) => {
+                        tracing::debug!(
+                            settlement_1 = ?settlement.solutions(),
+                            settlement_2 = ?other.solutions(),
+                            "merged settlements"
+                        );
+                        *other = m;
+                        merged = true;
+                        break;
+                    }
+                    Err(err) => {
+                        tracing::debug!(
+                            err = ?err,
+                            settlement_1 = ?settlement.solutions(),
+                            settlement_2 = ?other.solutions(),
+                            "settlements can't be merged"
+                        );
+                    }
+                }
+            }
+            // If [`settlement`] can't be merged into any other settlement, this is its
+            // final, most optimal form. Push it to the results.
+            if !merged {
+                results.push(settlement);
```

Comment:
nit: I think it might be easier to read to drop `merged` and instead move the push to the `break` statement.

Reply:
That's not the same thing. The `push` needs to happen if _no_ merge happened. OTOH what you're proposing would result in `push` happening when a merge happens, which is the opposite from what we want.

Reply:
Hmm, odd that this confused you. I thought the comments would make this really easy to follow. Am I missing something?

Reply:
Yeah, you are right. I remember stumbling over the same thing the first time I tried to review the PR and came to the conclusion that my suggestion is not equivalent. When I wanted to continue the review after github stopped having issues I tripped over the same thing but forgot about that insight. 🤦‍♂️